### ISSUE TITLE
v0.8.0 - created shape materials attribute is inaccessible / has a memory leak

### ISSUE THREAD

**Issue by Andrej730** [IfcOpenShell/IfcOpenShell#4833]
Have an issue trying to get shape's materials in v0.8.0, see example below (used file is - [cube_test.ifc.txt](https://github.com/user-attachments/files/15744266/cube_test.ifc.txt)). Build is from https://github.com/IfcOpenBot/IfcOpenShell/commit/90ae7091563348c42623220ea914c42d70485c03

@aothms can you please take a look?

```python
import ifcopenshell
import ifcopenshell.geom

path = r"cube_test.ifc"
ifc_file = ifcopenshell.open(path)

settings = ifcopenshell.geom.settings()
element = ifc_file.by_id(100)
shape = ifcopenshell.geom.create_shape(settings, element)
# <ifcopenshell.ifcopenshell_wrapper.Triangulation; proxy of <Swig Object of type 'IfcGeom::Representation::Triangulation *' at 0x000001A893AB8930> >
print(shape)
# (<Swig Object of type 'ifcopenshell::geometry::taxonomy::style *' at 0x0000015658A189F0>,)
# swig/python detected a memory leak of type 'ifcopenshell::geometry::taxonomy::style *', no destructor found.
print(shape.materials)
# Traceback (most recent call last):
#   File "test.py", line 16, in <module>
#     print([m.instance_id() for m in shape.materials])
#           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#   File "test.py", line 16, in <listcomp>
#     print([m.instance_id() for m in shape.materials])
#            ^^^^^^^^^^^^^
# AttributeError: 'SwigPyObject' object has no attribute 'instance_id'
print([m.instance_id() for m in shape.materials])
```

#4832

**Comment by Moult** [comment]
Confirming I get this too.

**Comment by Andrej730** [comment]
It seems this issue occur only with `settings.APPLY_DEFAULT_MATERIALS` set to `True` and in v0.8.0 it's set to `True` by default (by default it was `False` in v0.7.0).

**Comment by aothms** [comment]
Should be ok now. These are exactly the kind of architectural things that are conceptually small, but have big impacts in number of lines or missing functionality in the wrapper that we might see from time to time the coming month.

```
>>> import ifcopenshell, ifcopenshell.geom
No stream support: No module named 'lark'
>>> f = ifcopenshell.open('d.ifc')
>>> settings = ifcopenshell.geom.settings()
>>> element = f.by_type('IfcWall')[0]
>>> shape = ifcopenshell.geom.create_shape(settings, element)
>>> print(shape.geometry.materials[0])
style
     nameIfcSurfaceStyleRendering-17238
     diffuseIfcSurfaceStyleRendering-17238
            colour 0.501961 0.501961 0.501961
     specularIfcSurfaceStyleRendering-17238
             colour 0.00390625 0.00390625 0.00390625

```

It was a bit unfortunate timewise. Right when we made the switch I was caught in a bunch of workshops with the state of some things a bit inbetween. As you can see the style (coming from what we call the mapping) is now fully accessible. This will have some interesting possibilities in terms of how we interact with the geometry library. But we'll figure that out later.

**Comment by Moult** [comment]
> It seems this issue occur only with `settings.APPLY_DEFAULT_MATERIALS` set to `True` and in v0.8.0 it's set to `True` by default (by default it was `False` in v0.7.0).

Is this default change intentional? Just something we should watch out for.

**Comment by aothms** [comment]
> Is this default change intentional? Just something we should watch out for.

I guess partly. It was a bit messy and still is. Setting this to false breaks the serializers, because they cannot (out of pure utter laziness) cannot handle the -1 indices. I think for most users settings to the true is the sane default if they are not really informed, but I don't mind changing it to false either.

**Comment by Moult** [comment]
No worries, happy to keep the new default at True. We'll set it to False most likely in BBIM.

@Andrej730 can this be closed?

**Comment by Andrej730** [comment]
noticed another issue in ifc4/ifc2x3 - instance_id is returning IFCPRESENTATIONSTYLEASSIGNMENT's id instead of the surface style's id. Submitted a fix I was using in #4851

```python
import ifcopenshell
import ifcopenshell.geom

ifc_str = """
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition[DesignTransferView]'),'2;1');
FILE_NAME('style_ifc2x3.ifc','2024-06-12T13:30:48+05:00',(),(),'IfcOpenShell 0.0.0','BlenderBIM 0.0.999999-8241b55','Nobody');
FILE_SCHEMA(('IFC2X3'));
ENDSEC;
DATA;
#1=IFCPERSON('HSeldon','Seldon','Hari',$,$,$,$,$);
#2=IFCORGANIZATION('APTR','Aperture Science',$,$,$);
#3=IFCPERSONANDORGANIZATION(#1,#2,$);
#4=IFCACTORROLE(.USERDEFINED.,'CONTRIBUTOR',$);
#5=IFCTELECOMADDRESS(.USERDEFINED.,$,'WEBPAGE',$,$,$,$,'https://ifcopenshell.org');
#6=IFCORGANIZATION('IfcOpenShell','IfcOpenShell','IfcOpenShell is an open source software library that helps users and software developers to work with IFC data.',(#4),(#5));
#7=IFCAPPLICATION(#6,'0.0.999999','BlenderBIM Add-on','BlenderBIM');
#8=IFCOWNERHISTORY(#3,#7,.READWRITE.,.ADDED.,1718174537,#3,#7,1718174537);
#9=IFCPROJECT('23F$kwlK54SOSNlfvY9WV2',#8,'My Project',$,$,$,$,(#22,#34),#17);
#10=IFCSIUNIT(*,.LENGTHUNIT.,.MILLI.,.METRE.);
#11=IFCSIUNIT(*,.AREAUNIT.,$,.SQUARE_METRE.);
#12=IFCSIUNIT(*,.VOLUMEUNIT.,$,.CUBIC_METRE.);
#13=IFCDIMENSIONALEXPONENTS(0,0,0,0,0,0,0);
#14=IFCSIUNIT(*,.PLANEANGLEUNIT.,$,.RADIAN.);
#15=IFCMEASUREWITHUNIT(IFCREAL(0.0174532925199433),#14);
#16=IFCCONVERSIONBASEDUNIT(#13,.PLANEANGLEUNIT.,'degree',#15);
#17=IFCUNITASSIGNMENT((#12,#10,#16,#11));
#18=IFCCARTESIANPOINT((0.,0.,0.));
#19=IFCDIRECTION((0.,0.,1.));
#20=IFCDIRECTION((1.,0.,0.));
#21=IFCAXIS2PLACEMENT3D(#18,#19,#20);
#22=IFCGEOMETRICREPRESENTATIONCONTEXT($,'Model',3,1.E-05,#21,$);
#23=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body','Model',*,*,*,*,#22,$,.MODEL_VIEW.,$);
#24=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Axis','Model',*,*,*,*,#22,$,.GRAPH_VIEW.,$);
#25=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Box','Model',*,*,*,*,#22,$,.MODEL_VIEW.,$);
#26=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Annotation','Model',*,*,*,*,#22,$,.SECTION_VIEW.,$);
#27=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Annotation','Model',*,*,*,*,#22,$,.ELEVATION_VIEW.,$);
#28=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Annotation','Model',*,*,*,*,#22,$,.MODEL_VIEW.,$);
#29=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Annotation','Model',*,*,*,*,#22,$,.PLAN_VIEW.,$);
#30=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Profile','Model',*,*,*,*,#22,$,.ELEVATION_VIEW.,$);
#31=IFCCARTESIANPOINT((0.,0.));
#32=IFCDIRECTION((1.,0.));
#33=IFCAXIS2PLACEMENT2D(#31,#32);
#34=IFCGEOMETRICREPRESENTATIONCONTEXT($,'Plan',2,1.E-05,#33,$);
#35=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Axis','Plan',*,*,*,*,#34,$,.GRAPH_VIEW.,$);
#36=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body','Plan',*,*,*,*,#34,$,.PLAN_VIEW.,$);
#37=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Annotation','Plan',*,*,*,*,#34,$,.PLAN_VIEW.,$);
#38=IFCOWNERHISTORY(#3,#7,.READWRITE.,.MODIFIED.,1718174537,#3,#7,1718174537);
#39=IFCSITE('3O$YNBXT1FgxP14YgZtIsw',#38,'My Site',$,$,#65,$,$,.ELEMENT.,$,$,$,$,$);
#45=IFCOWNERHISTORY(#3,#7,.READWRITE.,.MODIFIED.,1718174537,#3,#7,1718174537);
#46=IFCBUILDING('3pjTJ88Qv0e8uwmCSVCpqP',#45,'My Building',$,$,#72,$,$,.ELEMENT.,$,$,$);
#52=IFCOWNERHISTORY(#3,#7,.READWRITE.,.MODIFIED.,1718174537,#3,#7,1718174537);
#53=IFCBUILDINGSTOREY('3IErWbOU19OPiga5hug7ZR',#52,'My Storey',$,$,#79,$,$,.ELEMENT.,$);
#59=IFCOWNERHISTORY(#3,#7,.READWRITE.,.ADDED.,1718174537,#3,#7,1718174537);
#60=IFCRELAGGREGATES('1ZyNXlPeHAKRcCUIiqjSJS',#59,$,$,#9,(#39));
#61=IFCCARTESIANPOINT((0.,0.,0.));
#62=IFCDIRECTION((0.,0.,1.));
#63=IFCDIRECTION((1.,0.,0.));
#64=IFCAXIS2PLACEMENT3D(#61,#62,#63);
#65=IFCLOCALPLACEMENT($,#64);
#66=IFCOWNERHISTORY(#3,#7,.READWRITE.,.ADDED.,1718174537,#3,#7,1718174537);
#67=IFCRELAGGREGATES('2413AEyGL3aBUiqqol6OgK',#66,$,$,#39,(#46));
#68=IFCCARTESIANPOINT((0.,0.,0.));
#69=IFCDIRECTION((0.,0.,1.));
#70=IFCDIRECTION((1.,0.,0.));
#71=IFCAXIS2PLACEMENT3D(#68,#69,#70);
#72=IFCLOCALPLACEMENT(#65,#71);
#73=IFCOWNERHISTORY(#3,#7,.READWRITE.,.ADDED.,1718174537,#3,#7,1718174537);
#74=IFCRELAGGREGATES('1nYLmdE_jAZg3bP5egMATg',#73,$,$,#46,(#53));
#75=IFCCARTESIANPOINT((0.,0.,0.));
#76=IFCDIRECTION((0.,0.,1.));
#77=IFCDIRECTION((1.,0.,0.));
#78=IFCAXIS2PLACEMENT3D(#75,#76,#77);
#79=IFCLOCALPLACEMENT(#72,#78);
#80=IFCOWNERHISTORY(#3,#7,.READWRITE.,.MODIFIED.,1718174542,#3,#7,1718174542);
#81=IFCBEAM('0Zin0tKq5EGvK_crzi6eSB',#80,'Cube',$,$,#126,#119,$);
#87=IFCCARTESIANPOINT((-999.999938964844,-999.999938964844,-999.999938964844));
#88=IFCCARTESIANPOINT((-999.999938964844,-999.999938964844,999.999938964844));
#89=IFCCARTESIANPOINT((-999.999938964844,999.999938964844,-999.999938964844));
#90=IFCCARTESIANPOINT((-999.999938964844,999.999938964844,999.999938964844));
#91=IFCCARTESIANPOINT((999.999938964844,-999.999938964844,-999.999938964844));
#92=IFCCARTESIANPOINT((999.999938964844,-999.999938964844,999.999938964844));
#93=IFCCARTESIANPOINT((999.999938964844,999.999938964844,-999.999938964844));
#94=IFCCARTESIANPOINT((999.999938964844,999.999938964844,999.999938964844));
#95=IFCPOLYLOOP((#87,#88,#90,#89));
#96=IFCFACEOUTERBOUND(#95,.T.);
#97=IFCFACE((#96));
#98=IFCPOLYLOOP((#89,#90,#94,#93));
#99=IFCFACEOUTERBOUND(#98,.T.);
#100=IFCFACE((#99));
#101=IFCPOLYLOOP((#93,#94,#92,#91));
#102=IFCFACEOUTERBOUND(#101,.T.);
#103=IFCFACE((#102));
#104=IFCPOLYLOOP((#91,#92,#88,#87));
#105=IFCFACEOUTERBOUND(#104,.T.);
#106=IFCFACE((#105));
#107=IFCPOLYLOOP((#89,#93,#91,#87));
#108=IFCFACEOUTERBOUND(#107,.T.);
#109=IFCFACE((#108));
#110=IFCPOLYLOOP((#94,#90,#88,#92));
#111=IFCFACEOUTERBOUND(#110,.T.);
#112=IFCFACE((#111));
#113=IFCCLOSEDSHELL((#97,#100,#103,#106,#109,#112));
#114=IFCFACETEDBREP(#113);
#115=IFCSHAPEREPRESENTATION(#23,'Body','Brep',(#114));
#116=IFCCARTESIANPOINT((-1000.,-1000.,-1000.));
#117=IFCBOUNDINGBOX(#116,2000.,2000.,2000.);
#118=IFCSHAPEREPRESENTATION(#25,'Box','BoundingBox',(#117));
#119=IFCPRODUCTDEFINITIONSHAPE($,$,(#118,#115));
#120=IFCOWNERHISTORY(#3,#7,.READWRITE.,.ADDED.,1718174542,#3,#7,1718174542);
#121=IFCRELCONTAINEDINSPATIALSTRUCTURE('0SxvH0z1T0oPwGwAclK6qW',#120,$,$,(#81),#53);
#122=IFCCARTESIANPOINT((0.,0.,0.));
#123=IFCDIRECTION((0.,0.,1.));
#124=IFCDIRECTION((1.,0.,0.));
#125=IFCAXIS2PLACEMENT3D(#122,#123,#124);
#126=IFCLOCALPLACEMENT(#79,#125);
#127=IFCSURFACESTYLE('red',.BOTH.,(#128));
#128=IFCSURFACESTYLESHADING(#129);
#129=IFCCOLOURRGB($,1.,0.,0.00230839615687728);
#133=IFCPRESENTATIONSTYLEASSIGNMENT((#127));
#134=IFCSTYLEDITEM(#114,(#133),$);
ENDSEC;
END-ISO-10303-21;
"""
ifc_file = ifcopenshell.file.from_string(ifc_str)

settings = ifcopenshell.geom.settings()
print(settings.set(settings.APPLY_DEFAULT_MATERIALS, False))
element = ifc_file.by_type("IfcBeam")[0]
representation = ifc_file.by_id(115)

import ifcopenshell.ifcopenshell_wrapper
ifcopenshell.ifcopenshell_wrapper.turn_on_detailed_logging()


shape = ifcopenshell.geom.create_shape(settings, element)

print(shape)
# v0.8.0:
# ['IfcSurfaceStyleShading-128']
# [133]

print([m.name for m in shape.geometry.materials])
print([m.instance_id() for m in shape.geometry.materials])
```
